File: leetcode-js/200-number-of-islands.js
/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function(grid) {
    if (!grid || !grid.length || !grid[0].length) {
        return 0;
    }

    let count = 0;

    for (let y = 0; y < grid.length; y += 1) {
        for (let x = 0; x < grid[0].length; x += 1) {
            if (grid[y][x] === '1') {
                count += 1;
                // Set all connected islands to 0
                merge(grid, x, y);
            }
        }
    }

    return count;
};

function merge(grid, x, y) {
    if (grid[y] == null || grid[y][x] == null || grid[y][x] === '0') {
        return;
    }

    grid[y][x] = '0';

    merge(grid, x - 1, y);
    merge(grid, x + 1, y);
    merge(grid, x, y - 1);
    merge(grid, x, y + 1);
}

File: leetcode-java/200-number-of-islands.java
public class Solution {
    public int numIslands(char[][] grid) {
        if (grid.length == 0) {
            return 0;
        }

        int m = grid.length;
        int n = grid[0].length;
        int count = 0;

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == '1') {
                    count++;
                    merge(grid, m, n, i, j);
                }
            }
        }

        return count;
    }

    void merge(char[][] grid, int m, int n, int i, int j) {
        if (i < 0 || i >= m || j < 0 || j >= n || grid[i][j] == '0') {
            return;
        }

        grid[i][j] = '0';

        merge(grid, m, n, i + 1, j);
        merge(grid, m, n, i - 1, j);
        merge(grid, m, n, i, j + 1);
        merge(grid, m, n, i, j - 1);
    }
}
